kb/2022-03-10/multiple_nics_vm_connectivity.md

```diff
@@ -0,0 +1,119 @@
+---
+title: Multiple NICs VM Connectivity
+descripion: How to deal VMs with multiple NICs in Harvester
+slug: multiple-nics-vm-connectivity
+authors:
+  - name: Date Huang
+    title: Software Engineer
+    url: https://github.com/tjjh89017
+    image_url: https://github.com/tjjh89017.png
+tags: [vm, network]
+hide_table_of_contents: false
+---
+
+# Multiple NICs VM Connectivity
+
+## What will happen in VM with multiple NICs by default in Harvester
+
+In some scenarios, we will setup two or more NICs for VM network to serve the different purpose. If setting up DHCP server and client by default in all networks, we might get connectivity issue randomly in VM. After some reboot, it might come back with expected connection, but it still will lost connection randomly after some period.
+
+## Identify your connectivity issue that is covered in this knowledge base
+
+We take Linux VM and `iproute2` as example to identify your default route. If you cannot execute the commnad below, please switch root user or execute via `sudo` and execute again.
+
+In your VM, execute following command in your VM in normal state:
+```bash
+ip route show default
+```
+
+In normal state, you will get only one route with destation to default and show the correct gateway IP.
+```
+default via <Gateway IP> dev eth0 proto dhcp src <VM IP> metric 100
+```
+
+For example:
+```
+$ ip route show default
+default via 192.168.0.254 dev eth0 proto dhcp src 192.168.0.100 metric 100
+$
+```
+
+If issue that covered by this KB occurs, you could login VM via Web Console in Harvester Web UI and execute following command in your abnormal VM:
+```bash
+ip route show default
+```
+
+In abnormal state, you will get one default route with incorrect gateway IP.
+For example:
+```
+default via <Incorrect Gateway IP> dev eth0 proto dhcp src <VM's IP> metric 100
+```
+
+## Why cause connectivity issue randomly in VM.
+
+In most of time, we will enable cloud based VM with DHCP on all NICs and enable DHCP server with gateway IP. When DHCP client receives response and gets gateway IP, it will setup IP address and default route to the gateway IP, so users can access VM via SSH to the IP.
+
+Mostly Linux distribution will treat multiple NICs without any dependency and allow only one default route with same distance by default. It will start multiple DHCP clients simultaneously. If we have two NICs with DHCP and both have gateway IP in DHCP response, two DHCP client will race for setting default route to Linux network. According to different implementation of DHCP script, former and latter will not always win to setup default route. This will cause unstable default route in every DHCP renew process or machine boot up. VM's connection will be unstablei default route is always changing.
```

Comment:
This will cause unstable default route in every DHCP renew process or machine boot up.
--> This will cause an unstable default route in every DHCP renewing process or machine boot up.

VM's connection will be unstablei default route is always changing.
--> VM's connection will be unstable if the default route is changing.

Reply:
```suggestion
Mostly Linux distribution will treat multiple NICs without any dependency and allow only one default route with the same distance by default. It will start multiple DHCP clients simultaneously. If we have two NICs with DHCP and both have gateway IP in DHCP response, two DHCP clients will race for setting default route to Linux network. According to different implementations of DHCP script, the former and latter will not always win to set up the default route. This will cause an unstable default route in every DHCP renewing process or machine boot up. VM's connection will be unstable if the default route is changing.
```